percy-mirror-verifier/src/server.ts

```diff
@@ -0,0 +1,154 @@
+import * as github from './github';
+import * as percy from './percy';
+import * as webhooks from './webhooks';
+import {PercyWebhook} from './webhooks';
+
+export function verify(
+  req: IncomingMessage,
+  res: ServerResponse,
+  buffer: Buffer
+) {
+  const expectedDigest = crypto
+    .createHmac('sha256', process.env.PERCY_WEBHOOK_SECRET)
+    .update(buffer)
+    .digest('hex');
+  if (req.headers['x-percy-digest'] !== expectedDigest) {
+    throw new Error('Message digest is incorrect for PERCY_WEBHOOK_SECRET');
+  }
+}
+
+export async function handleBuildFinished(
+  included: webhooks.PercyWebhookIncluded[]
+): Promise<void> {
+  const percyMainBuild = webhooks.getIncluded(included, 'builds');
+  if (percyMainBuild.attributes.branch !== 'main') {
+    console.log('This build is not a main branch build');
+    return;
+  }
+
+  try {
+    const percyMainBuildId = webhooks.getBuildId(included);
+    console.log('Percy build id for main branch:', percyMainBuildId);
+    const githubPullNumber = webhooks.getPullNumber(included);
+    console.log(
+      'Pull request that was merged into this commit:',
+      githubPullNumber
+    );
+
+    const percyPullBuildId = await github.getPercyBuildId(githubPullNumber);
+    console.log('Percy build id for the pull request:', percyPullBuildId);
+
+    const prSnapshots = await percy.getSnapshots(percyPullBuildId);
+    const mainSnapshots = await percy.getSnapshots(percyMainBuildId);
+
+    if (prSnapshots.size === 1 && prSnapshots.has('Blank page')) {
+      console.log(
+        'Pull request',
+        githubPullNumber,
+        'was a docs-only change. Skipping...'
+      );
+      return;
+    }
+
+    // We ignore the Blank page.
+    mainSnapshots.delete('Blank page');
+
+    // Mirroring verification: the following must all be true to indicate a
+    // correct mirroring of the PR/main builds:
+    // * The number of snapshots on both the PR and main build is equal (after
+    //   we remove the "Blank page" snapshot)
+    // * The set of snapshot names are equal
+    // * All snapshots are approved
+    // * All snapshots must have the same approval reason (both are no_diff or
+    //   were both approved, manually/automatically)
+    // If any of these are false, this indicates a disparity between the PR and
+    // main builds.
+    if (
+      prSnapshots.size != mainSnapshots.size ||
+      [...prSnapshots.keys()].some(name => !mainSnapshots.has(name))
+    ) {
+      // TODO(danielrozenberg): write a comment on the PR (here and below).
+      console.error('PR/main disparity: not all snapshots exist in both');
+      return await github.postErrorComment(githubPullNumber);
+    }
+
+    if (
+      [...prSnapshots.values(), ...mainSnapshots.values()].some(
+        ({attributes}) => attributes['review-state'] !== 'approved'
+      )
+    ) {
+      console.error('PR/main disparity: not all snapshots are marked approved');
+      return await github.postErrorComment(githubPullNumber);
```

Comment:
post-submit suggestion: add a `message` param to `postErrorComment` since most users/PR authors won't have access to the logs (I know the method isn't implemented yet, but when you do)

Reply:
This will just post an error comment to on the pull request. I didn't bother making this generic because (atm) I don't see this app getting complex enough to merit that